[slider][material-next] Slider restructure and style improvements

DiegoAndai opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

**Material Next Slider issue**: https://github.com/mui/material-ui/issues/37527

## Summary

This PR implements improvements that required a restructuring of the Slider Thumb and Value Label structure. The original PR introducing `material-next`'s Slider is https://github.com/mui/material-ui/pull/37520

The improvements are:

### Thumb restructure

Address [this comment](https://github.com/mui/material-ui/blob/f74848f86f855d5767732ecc91a71629a548bb18/packages/mui-material-next/src/Slider/Slider.tsx#L755). Restructure the Thumb and value label to a simpler structure, like the one in [Joy's Slider](https://github.com/mui/material-ui/blob/master/packages/mui-joy/src/Slider/Slider.tsx#L610). This removes the need for a separate ValueLabel component and types file, so those were removed.

### Thumb and value label overlap

Implement the [overlap border specification](https://m3.material.io/components/sliders/guidelines#ad5ceb95-a690-4ddd-8243-53a8e13bdab6) from Material You.

To do so I introduced the `useSliderElementsOverlap` hook that handles the logic of detecting overlaps for thumbs and value labels.

https://github.com/mui/material-ui/assets/16889233/d5c7d262-91e5-4286-9263-012b5a52f35f

DiegoAndai committed:
Restructure slider thumb, input, and value label

DiegoAndai committed:
Implement Slider overlap borders

DiegoAndai committed:
Stop forwarding value label display/format props

DiegoAndai committed:
Add slider thumb and label overlap tests

DiegoAndai committed:
Fix prettier issues

DiegoAndai committed:
Fix skip flag incorrect placing

DiegoAndai committed:
Fix proptypes commited test step

mnajdova commented on packages/mui-material-next/src/Slider/Slider.test.js:
```diff
@@ -1437,4 +1441,48 @@ describe('<Slider />', () => {
+
+  describe('overlapping state', () => {
+    it('should apply thumb overlap classes', function test() {
+      if (/jsdom/.test(window.navigator.userAgent)) {
+        this.skip();
+      }
```
```suggestion
```

I don't see why we should skip this in JSDOM. If it is not necessary let's remove it from all test, it will make the tests run faster.

DiegoAndai replied:
It's "necessary" because `getBoundingClientRect()` always returns `0, 0, 0, 0` in JsDOM (https://github.com/jsdom/jsdom/issues/653) and the element overlap calculation relies on it. In Karma, the tests work fine as `getBoundingClientRect()` returns the correct position.

What I could do is stub `getBoundingClientRect()`. Do you think that's better? I think the test is stronger without stubbing, but I would like to get your opinion.

mnajdova replied:
Ah I see, let's keep it as is, thanks for the clarification.

mnajdova commented on packages/mui-material-next/src/Slider/useSliderElementsOverlap.ts:
```diff
@@ -0,0 +1,87 @@
+import { unstable_useEnhancedEffect as useEnhancedEffect } from '@mui/utils';
+
+function getSliderElementsOverlap(elements: HTMLElement[], axis: Axis, margin: number) {
+  if (elements.length === 2) {
```
if we accept an array, let's make the hook considering all elements in the array, it shouldn't be much change.

DiegoAndai replied:
I agree! But rather than considering all elements, I modified it to only accept two at most. I think this makes more sense as the slider thumb and value label elements should be two at most as well.

mnajdova reviewed (commented):
Overall it looks great! Left some comments for improvements.

DiegoAndai committed:
Restrain useSliderElementsOverlap to 2 elements

DiegoAndai committed:
Add missing Slider export from material-next index

DiegoAndai committed:
Use common token for white instead of string

DiegoAndai committed:
Add v6 component changes page

DiegoAndai commented:
Hey @mnajdova, thanks for the review! I addressed your comments and also I added the v6 components migration guide, that way we can begin documenting breaking changes alongside PRs, with this PR introducing the first.

mnajdova commented on docs/data/material/migration/migration-v6/v6-component-changes.md:
```diff
@@ -0,0 +1,38 @@
+# Breaking changes in v6: core components
```
Instead of page, can we please add readme file in the @mui/material-next package? Maybe with name `migration.md` next to readme. We can add info in the readme, that they should follow the `migraiton.md` guidelines.

mnajdova reviewed (commented):
The changes look good to me, I left one suggestion for the migration location, as we anyway don't want to add a page on the docs just yet. @DavidCnoops can you review the design-related changes?

DavidCnoops reviewed (approved):
Overall this looks really good to me! It's a great solution to a tricky UI problem.

I have one very tiny nitpick: when valueLabelDisplay is set to `auto` and the 2 thumbs/values are about to overlap, the underlying value very briefly pops up. Also, when sliding the upper thumb over the lower one (i.e. when they completely overlap) its value seems to briefly disappear. It's -extremely- subtle so definitely not a big or blocking issue, but a 100% perfect solution would be to keep the lower value constantly hidden and the upper value constantly visible.

[slider.webm](https://github.com/mui/material-ui/assets/28001064/71e46b83-04f4-453a-b9a6-d9f681cd86ee)

DiegoAndai committed:
Move migration guide from Docs to GitHub

DiegoAndai committed:
Use Material You as preferred name

DiegoAndai committed:
Fix value label overlap animation issues

DiegoAndai commented:
Hey @DavidCnoops! thanks for reviewing 😊 

What you reported was indeed a bug:

- The active thumb's `z-index` should be higher than the other thumb to avoid the inactive one triggering hover states when dragging the active one, this fixes the "about to overlap issue"
- The `transform` animation should be disabled for the value label when dragging the thumb, this fixes the "briefly disappear when overlapping" issue

Thanks for catching the issues 🚀

mnajdova reviewed (commented):
We have a demo of a Slider with three thumbs - http://localhost:3000/material-ui/react-slider/#removed-track, also as far as I remember we updated the Base UI Slider and useSlider to support more than two thumbs. I would propose making this scenario works too.

DiegoAndai commented on packages/mui-material-next/src/Slider/useSliderElementsOverlap.ts:
```diff
@@ -0,0 +1,92 @@
+
+function useElementsOverlap(axis: Axis, margin: number = 0) {
+  const elementsRefList = React.useRef<SliderElements>([]);
+  const [elementsOverlap, setThumbsOverlap] = React.useState(false);
```
Nice catch 😄

DiegoAndai commented on packages/mui-material-next/src/Slider/useSliderElementsOverlap.ts:
```diff
@@ -0,0 +1,92 @@
+    getIsOverlapping: getIsValueLabelOverlapping,
+    onMove: onValueLabelMove,
+  } = useElementsOverlap(axis, -12);
+  // ValueLabel -12px margin is required due to how its "inverted water drop"
+  // shape is built with CSS. Might want to allow this to be configurable
+  // in the future.
```
I put it here:

```js
  // ValueLabel -12px margin is required due to how its "inverted water drop"
  // shape is built with CSS. Might want to allow this to be configurable in the future.
  const {
    setRef: setValueLabelRef,
    getIsOverlapping: getIsValueLabelOverlapping,
    onMove: onValueLabelMove,
  } = useElementsOverlap(axis, -12);
```

The other option was

```js
  const {
    setRef: setValueLabelRef,
    getIsOverlapping: getIsValueLabelOverlapping,
    onMove: onValueLabelMove,
    // ValueLabel -12px margin is required due to how its "inverted water drop"
    // shape is built with CSS. Might want to allow this to be configurable in the future.
  } = useElementsOverlap(axis, -12);
```

But I didn't like that one, which one do you prefer?

DiegoAndai committed:
Support multiple elements and refactor performance

DiegoAndai commented:
Hey @mnajdova thanks for the feedback, I included your changes in the last commit:

- Improve naming and code structure in the `useSliderElementsOverlap.ts` file
- Support more than 2 values for the overlap style
- Added debounce function to `onMove` callback to reduce unnecessary overlap calculation. I tested and a debounce of `50ms` felt the best UXwise while also reducing the calculations considerably.

mnajdova reviewed (approved):
Looks much better 👍 I would recommend for PRs that include fixes like this to share a codesandbox for easily verifying the fix. It would help a lot with the reviews :)
